### PR TITLE
Add analysis notebook

### PR DESCRIPTION
**Pull request by hideaki-j**
Add `analysis.ipynb` for dataset analysis (Issue #3)

**Details:**
- `analysis.ipynb` includes all your analyses + correlation and diversity analysis.
- `metrics.py` contains code for Distinct-n calculation.
- Upon reviewing the analysis code, I found 2 invalid votes (from users without any conversations stored); those are removed from `votes_open.csv`.

### CODE REVIEW

File: tool/metrics.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
+"""
+Implements the Distinct-n metric for evaluating diversity in dialogues.
+The implementation is copied from Joko et al. (2024) which is based on Tevet et al. (2021).
+For more details, refer to Joko et al. (2024).
+
+References:
+- Joko et al. (2024): https://dl.acm.org/doi/10.1145/3626772.3657815
+    (GitHub: https://github.com/informagi/laps)
+- Tevet et al. (2021): https://aclanthology.org/2021.eacl-main.25/
```
Comment (NoB0) [inline tool/metrics.py]: Nit: remove empty first line.

```suggestion
"""Implements the Distinct-n metric for evaluating diversity in dialogues.
The implementation is copied from Joko et al. (2024) which is based on Tevet et al. (2021).
For more details, refer to Joko et al. (2024).

References:
- Joko et al. (2024): https://dl.acm.org/doi/10.1145/3626772.3657815
    (GitHub: https://github.com/informagi/laps)
- Tevet et al. (2021): https://aclanthology.org/2021.eacl-main.25/
```

File: tool/metrics.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+"""
+
+import random
+from nltk.tokenize import word_tokenize
+import numpy as np
```
Comment (NoB0) [inline tool/metrics.py]: Nit: looks like numpy is not used.

File: tool/metrics.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        self.use_lower_case = use_lower_case
+
+    def _tokenize(self, line: str) -> List[str]:
+        """Tokenizes the input line based on the specified tokenizer_name.
+
+        Args:
+        line (str): The input line to _tokenize.
+
+        Returns:
+        list: A list of tokens from the input line.
```
Comment (NoB0) [inline tool/metrics.py]: Nit: add indentation for consistency.

```suggestion
        Args:
            line (str): The input line to _tokenize.

        Returns:
            list: A list of tokens from the input line.
```

File: tool/metrics.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    def _calculate_unique_ngram_ratio(self, ngram_lists: List[List[Tuple[str, ...]]]) -> float:
+        ngrams = [item for sublist in ngram_lists for item in sublist]  # flatten
+        return len(set(ngrams)) / len(ngrams) if len(ngrams) > 0 else 0.
+
+    def calculate_distinct_n(self, response_set: List[str]) -> float:
+        return self._calculate_unique_ngram_ratio(self.lines_to_ngrams(response_set))
+
+    def calculate_normalized_distinct_n(self, response_set: List[str], return_avg: bool = True, max_token_length: int = 7012, iterations: int = 100, seed: int = 42) -> Union[float, List[float]]:
+        """Calculate the truncated Distinct-n metric for a set of responses.
```
Comment (NoB0) [inline tool/metrics.py]: Nit: 
```suggestion
        """Calculates the truncated Distinct-n metric for a set of responses.
```

**Review by NoB0 — COMMENTED** [review]
A few comments to consider but mostly details. I suggest to use black formatter to improve readability of the code.

File: tool/metrics.py (hunk 2 above)
Comment (hideaki-j) [reply tool/metrics.py]: Done.

**Comment by hideaki-j** [comment]
Thanks for the review - I have applied your suggestions, please confirm them.

**Comment by NoB0** [comment]
I do not see changes in the notebook, please check the comments above.

**Comment by hideaki-j** [comment]
<p>Added <code>!pip install</code> block before imports.</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785715'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>Removed <code>import itertools</code> </p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785742'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>Renamed "Sentiment analysys" to "User satisfaction analysis"</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785768'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>Replaced all <code>sentiment_</code> with  <code>satisfaction_</code> in the notebook.</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785783'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>Moved "Closed vs. open crowdsourcing" under  "Conversation" (Conversation statistics).</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785854'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>removed feedback section and corresponding imports</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785829'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
<p>Replace "battle outcomes" by "pairwise comparison".</p><p> </p><p>Note that the term "battle" is used everwhere and I did not replace all with "pairwise comparison" to avoid any further inconsistency issues. (also the term battle is used in the paper so I think it is ok.)</p>

---

 <a href='https://app.reviewnb.com/iai-group/crsarena-dial/pull/4/discussion/#comment-2387785805'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by hideaki-j** [comment]
Updated the notebook according to your review. (Sorry for missing it; I did not have a ReviewNB account and your review was not shown properly).

**Review by NoB0 — APPROVED** [review]
LGTM.

**Pull request merged**